trust/p11-kit-verify.c

```diff
@@ -0,0 +1,392 @@
+ *       the following disclaimer in the documentation and/or
+ *       other materials provided with the distribution.
+ *     * The names of contributors to this software may not be
+ *       used to endorse or promote products derived from this
+ *       software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: Zoltan Fridrich <zfridric@redhat.com>
+ */
+
+#include <ctype.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define IS_BASE64(C) (isalnum((int)(C)) || (C) == '+' || (C) == '/')
+#define PEM_LABEL_MAX 256
+
+char pem_label[PEM_LABEL_MAX];
+int pem_label_length;
+int inside_pem;
+int inside_section;
+size_t line_num;
+
+static inline void fail(const char *str)
+{
+    fprintf(stderr, "Line %lu: %s\n", line_num, str);
+}
+
+static char *get_line(FILE *stream)
+{
+    char *buffer, *tmp;
+    int c, count, alloc_size;
+
+    count = 0;
+    alloc_size = 50;
+
+    buffer = malloc(alloc_size);
+    if (buffer == NULL) {
+        fprintf(stderr, "Error: failed to allocate memory\n");
+        return NULL;
+    }
+
+    while ((c = getc(stream)) != EOF && c != '\n') {
+        if (count + 1 >= alloc_size) {
+            alloc_size *= 2;
+            tmp = realloc(buffer, alloc_size);
+            if (tmp == NULL) {
+                free(buffer);
+                fprintf(stderr, "Error: failed to allocate memory\n");
+                return NULL;
+            }
+            buffer = tmp;
+        }
+        buffer[count++] = (char)c;
+    }
+
+    if (c == EOF)
+        return NULL;
+
+    buffer[count] = '\0';
+    return buffer;
+}
+
+/* Format is based on https://www.rfc-editor.org/rfc/rfc7468#section-3
+ *
+ * stricttextualmsg = preeb eol
+ *                    strictbase64text
+ *                    posteb eol
+ * preeb            = "-----BEGIN " label "-----"
+ * posteb           = "-----END " label "-----"
+ * strictbase64text = *base64fullline strictbase64finl
+ * base64fullline   = 64base64char eol
+ * strictbase64finl = *15(4base64char) (4base64char / 3base64char base64pad / 2base64char 2base64pad) eol
+ */
+static bool parse_pem(const char *line)
```

Comment:
Can't we use  `p11_pem_parse` from trust/pem.c?

Reply:
No.
There are several reasons why we don't want to use  `p11_pem_parse`.
1. This program is supposed to verify `.p11-kit` file format and report any errors to the user. If I were to use  `p11_pem_parse` there would be no usefull error messages for PEM format.
2. My version is more strict. For example `p11_pem_parse` does not seem to check whether the labels in PEM headers match.
3.  This program parses the file on per-line basis while `p11_pem_parse` doesn't.